@mjklemm, I made the required changes, I have added some more testcases, do you see anything missing there.
Also, I have a doubt, consider an example:
```fortran
program single
    !$omp single nowait
        print *, x
    !$omp end single copyprivate(x, y) nowait
end program
```
Error:
```console
$ flang x.f90 -fopenmp
flang-21: warning: OpenMP support in flang is still experimental [-Wexperimental-option]
error: Semantic errors in x.f90
./x.f90:2:18: error: NOWAIT clause must not be used with COPYPRIVATE clause on the SINGLE directive
      !$omp single nowait
                   ^^^^^^
./x.f90:4:40: warning: NOWAIT clause is already used on the SINGLE directive
      !$omp end single copyprivate(x, y) nowait
                                         ^^^^^^
```
Is the error and the warning correct?

--- a/flang/lib/Semantics/check-omp-structure.cpp
+++ b/flang/lib/Semantics/check-omp-structure.cpp
@@ -1856,8 +1856,8 @@ void OmpStructureChecker::Enter(const parser::OpenMPDispatchConstruct &x) {
 
   auto it{block.begin()};
   bool passChecks{false};
-  if (const parser::AssignmentStmt *assignStmt{
-          parser::Unwrap<parser::AssignmentStmt>(*it)}) {
+  if (const parser::AssignmentStmt *
+      assignStmt{parser::Unwrap<parser::AssignmentStmt>(*it)}) {
     if (parser::Unwrap<parser::FunctionReference>(assignStmt->t)) {
       passChecks = true;
     }

--- a/flang/lib/Semantics/check-omp-structure.cpp
+++ b/flang/lib/Semantics/check-omp-structure.cpp
@@ -1206,8 +1206,7 @@ void OmpStructureChecker::Enter(const parser::OpenMPBlockConstruct &x) {
   if (GetContext().directive == llvm::omp::Directive::OMPD_single) {
     std::set<Symbol *> singleCopyprivateSyms;
     std::set<Symbol *> endSingleCopyprivateSyms;
-    bool singleNowait{false};
-    bool endSingleNowait{false};
+    bool foundNowait{false};
     parser::CharBlock NowaitSource;
 
     auto catchCopyPrivateNowaitClauses = [&](const auto &dir, bool endDir) {
@@ -1240,23 +1239,11 @@ void OmpStructureChecker::Enter(const parser::OpenMPBlockConstruct &x) {
             }
           }
         } else if (clause.Id() == llvm::omp::Clause::OMPC_nowait) {
-          if (singleNowait) {
-            if (endDir) {
-              context_.Warn(common::UsageWarning::OpenMPUsage, clause.source,
-                  "NOWAIT clause is already used on the SINGLE directive"_warn_en_US);
-            } else {
-              context_.Say(clause.source,
-                  "At most one NOWAIT clause can appear on the SINGLE directive"_err_en_US);
-            }
-          } else if (endSingleNowait) {
+          if (foundNowait) {
             context_.Say(clause.source,
-                "At most one NOWAIT clause can appear on the END SINGLE directive"_err_en_US);
+                "At most one NOWAIT clause can appear on the SINGLE directive"_err_en_US);
           } else {
-            if (endDir) {
-              endSingleNowait = true;
-            } else {
-              singleNowait = true;
-            }
+            foundNowait = !endDir;
           }
           if (!NowaitSource.ToString().size()) {
             NowaitSource = clause.source;

--- a/flang/test/Semantics/OpenMP/clause-validity01.f90
+++ b/flang/test/Semantics/OpenMP/clause-validity01.f90
@@ -334,8 +334,8 @@ use omp_lib
   !$omp single private(a) lastprivate(c) nowait
   a = 3.14
   !ERROR: COPYPRIVATE variable 'a' may not appear on a PRIVATE or FIRSTPRIVATE clause on a SINGLE construct
-  !WARNING: NOWAIT clause is already used on the SINGLE directive
-  !WARNING: NOWAIT clause is already used on the SINGLE directive
+  !ERROR: At most one NOWAIT clause can appear on the SINGLE directive
+  !ERROR: At most one NOWAIT clause can appear on the SINGLE directive
   !ERROR: At most one NOWAIT clause can appear on the END SINGLE directive
   !$omp end single copyprivate(a) nowait nowait
   c = 2

--- a/flang/test/Semantics/OpenMP/single03.f90
+++ b/flang/test/Semantics/OpenMP/single03.f90
@@ -44,7 +44,7 @@ subroutine omp_single
 
         !$omp single nowait
             print *, "omp single", j
-        !WARNING: NOWAIT clause is already used on the SINGLE directive
+        !ERROR: At most one NOWAIT clause can appear on the SINGLE directive
         !$omp end single nowait
     !$omp end parallel
 

--- a/flang/test/Semantics/OpenMP/single04.f90
+++ b/flang/test/Semantics/OpenMP/single04.f90
@@ -53,7 +53,7 @@ program single
     !$omp single copyprivate(x) nowait
         print *, x
     !WARNING: The COPYPRIVATE clause with 'x' is already used on the SINGLE directive
-    !WARNING: NOWAIT clause is already used on the SINGLE directive
+    !ERROR: At most one NOWAIT clause can appear on the SINGLE directive
     !$omp end single copyprivate(x) nowait
 
     !$omp single copyprivate(x)
@@ -67,7 +67,7 @@ program single
         print *, x
     !WARNING: The COPYPRIVATE clause with 'x' is already used on the SINGLE directive
     !ERROR: 'z' appears in more than one COPYPRIVATE clause on the END SINGLE directive
-    !WARNING: NOWAIT clause is already used on the SINGLE directive
+    !ERROR: At most one NOWAIT clause can appear on the SINGLE directive
     !$omp end single copyprivate(x, z) copyprivate(z) nowait
 
     !ERROR: NOWAIT clause must not be used with COPYPRIVATE clause on the SINGLE directive
@@ -76,6 +76,6 @@ program single
     !WARNING: The COPYPRIVATE clause with 'x' is already used on the SINGLE directive
     !WARNING: The COPYPRIVATE clause with 'y' is already used on the SINGLE directive
     !WARNING: The COPYPRIVATE clause with 'z' is already used on the SINGLE directive
-    !WARNING: NOWAIT clause is already used on the SINGLE directive
+    !ERROR: At most one NOWAIT clause can appear on the SINGLE directive
     !$omp end single copyprivate(x, y, z) nowait
 end program

--- a/llvm/include/llvm/Frontend/OpenMP/OMP.td
+++ b/llvm/include/llvm/Frontend/OpenMP/OMP.td
@@ -998,8 +998,10 @@ def OMP_Single : Directive<"single"> {
     VersionedClause<OMPC_Allocate>,
     VersionedClause<OMPC_CopyPrivate>,
     VersionedClause<OMPC_FirstPrivate>,
-    VersionedClause<OMPC_NoWait>,
     VersionedClause<OMPC_Private>,
+  ];
+  let allowedOnceClauses = [
+    VersionedClause<OMPC_NoWait>,
   ];
   let association = AS_Block;
   let category = CA_Executable;

